pandera/engines/polars_engine.py

```diff
@@ -0,0 +1,496 @@
+    polars_failure_cases_from_coercible,
+    check_polars_container_all_true,
+)
+
+
+@immutable(init=True)
+class DataType(dtypes.DataType):
+    """Base `DataType` for boxing Polars data types."""
+
+    type: pl.DataType = dataclasses.field(repr=False, init=False)
+
+    def __init__(self, dtype: Optional[Any] = None):
+        super().__init__()
+        try:
+            object.__setattr__(self, "type", py_type_to_dtype(dtype))
+        except ValueError:
+            object.__setattr__(self, "type", pl.Object)
+
+        dtype_cls = dtype if inspect.isclass(dtype) else dtype.__class__
+        warnings.warn(
+            f"'{dtype_cls}' support is not guaranteed.\n"
+            + "Usage Tip: Consider writing a custom "
+            + "pandera.dtypes.DataType or opening an issue at "
+            + "https://github.com/pandera-dev/pandera"
+        )
+
+    def __post_init__(self):
+        # this method isn't called if __init__ is defined
+        object.__setattr__(
+            self, "type", py_type_to_dtype(self.type)
+        )  # pragma: no cover
+
+    def coerce(self, data_container: PolarsObject) -> PolarsObject:
+        """Coerce data container to the data type."""
+        return data_container.cast(self.type, strict=True)
+
+    def try_coerce(self, data_container: PolarsObject) -> PolarsObject:
+        """Coerce data container to the data type,
+        raises a :class:`~pandera.errors.ParserError` if the coercion fails
+        :raises: :class:`~pandera.errors.ParserError`: if coercion fails
+        """
+        try:
+            return self.coerce(data_container)
+        except Exception as exc:  # pylint:disable=broad-except
+            raise errors.ParserError(
+                f"Could not coerce {type(data_container)} data_container "
+                f"into type {self.type}",
+                failure_cases=polars_coerce_failure_cases(
+                    data_container=data_container, type_=self.type
+                ),
+            ) from exc
+
+    def check(
+        self,
+        pandera_dtype: dtypes.DataType,
+        data_container: Optional[PolarsObject] = None,
+    ) -> Union[bool, Iterable[bool]]:
+        try:
+            return self.type == pandera_dtype
+        except TypeError:  # pragma: no cover
+            return False
+
+    def __str__(self) -> str:
+        return str(self.type)
+
+    def __repr__(self) -> str:
+        return f"DataType({self})"
+
+
+class Engine(  # pylint:disable=too-few-public-methods
+    metaclass=engine.Engine, base_pandera_dtypes=DataType
+):
+    """Polars data type engine."""
+
+    @classmethod
+    def dtype(cls, data_type: Any) -> dtypes.DataType:
+        """Convert input into a polars-compatible
+        Pandera :class:`~pandera.dtypes.DataType` object."""
+        try:
+            return engine.Engine.dtype(cls, data_type)
+        except TypeError:
+            try:
+                pl_dtype = py_type_to_dtype(data_type)
+            except TypeError:
```

Comment:
Seems like in the latest Polars version '0.20.8', `py_type_to_dtype` throws a ValueError instead of a TypeError

Reply:
Actually, you are right that the error is wrong. It should catch a ValueError if I am not mistaken. Here is the snippet that raises the error from polars 0.20.8 `    try:
        return _map_py_type_to_dtype(data_type)
    except (KeyError, TypeError):  # pragma: no cover
        if not raise_unmatched:
            return None
        msg = f"cannot infer dtype from {data_type!r} (type: {type(data_type).__name__!r})"
        raise ValueError(msg) from None`